### PR TITLE
Go 1.18 Survey

### PR DESCRIPTION
**Pull request by CamberLoid**
Topic Description
-----------------

This topic will update `go` to ~~1.18~~ 1.18.1. Check updates and rebuild all packages of `groups/go-rebuilds`.

Package(s) Affected
-------------------

`go`
`groups/go-rebuilds`
`extra-base/golang-base`

And others
`scons` `flatbuffers`

cmdline: `ciel clean;ciel build -i {$instance} go go-md2man groups/go-rebuilds`

Security Update?
----------------

No. 

Build Order
-----------

0. `scons` Noarch dependency. build and push first;
1. `go` The compiler itself;
2. `golang-base` `go-md2man` Toolkits for go;
3. `groups/go-rebuilds` All other packages;
4. `flatbuffers` Non-go-based package.

Test Build(s) Done
------------------

The checklist: https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923#issuecomment-1120379096

**Primary Architectures**

- [x] AMD64 `amd64`   
- [x] AArch64 `arm64`
    

**Secondary Architectures**

Architectural progress for "secondary," or experimental ports does not impede on merging of this topic.

- [x] Loongson 3 `loongson3`
- [x] ~~PowerPC 64-bit (Little Endian) `ppc64el`~~
- [x] RISC-V 64-bit `riscv64`

Update(s) Uploaded to Stable
----------------------------

**Primary Architectures**

- [ ] AMD64 `amd64`   
- [ ] AArch64 `arm64`

**Secondary Architectures**

Architectural progress for "secondary," or experimental ports does not impede on merging of this topic.

- [ ] Loongson 3 `loongson3`
- [ ] ~~PowerPC 64-bit (Little Endian) `ppc64el`~~
- [ ] RISC-V 64-bit `riscv64`

### CODE REVIEW

**Comment by CamberLoid** [comment]
Upstream has released newer go compiler 1.18.1 . Previously updated packages will be rebuilt.

**Comment by CamberLoid** [comment]
`popub` is tested and perform well on `riscv64`.
Despite there is no major behavior change between two commit of upstream, this package may require further tests due to the change of compiler from gccgo to google-go

**Comment by CamberLoid** [comment]
Test checklist for `frp` | in $SRCDIR `(uname -a; HOME=/root make test) | curl -F "c=@-" "http://fars.ee/"`
- [x] amd64 | Log: http://fars.ee/CjeV
- [x] arm64 | Log: http://fars.ee/chDl
- [x] loongson3 | Log: http://fars.ee/eYJ4
- [x] ppc64el | http://fars.ee/vecw
- [x] riscv64 | Log: http://fars.ee/HvRr

**Comment by CamberLoid** [comment]
Test checklist for hugo | Can successfully host https://github.com/AOSC-Dev/aosc-portal-kiss.github.io , and try to access it via `curl -L` or `lynx` (modern browser if possible)
- [x] amd64 
image
- [x] arm64
image
- [x] ppc64el
image
- [x] loongson3
image
- [x] riscv64 
image

**Comment by CamberLoid** [comment]
Test checklist for packages, so far
1. Core component
- [x] The compiler itself `go` | As long as there is no building error for other packages, considering this is tested
2. Packages to upgrade
- [x] `delve` 
- [x] `frp` https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923#issuecomment-1120217271
- [x] `hugo` https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923#issuecomment-1120233524
- [x] `gotop` https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923#issuecomment-1120590282 | Functionality tests are required as this upgrade enables `-buildmode=pie`
- [ ] ~~`clash`~~ https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923#issuecomment-1121127189 | Skipped/Delayed
- [x] `popub` | Proved functionality on riscv64. 
- [ ] `git-lfs` https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923#issuecomment-1148260409
- [x] `gh`/`github-cli` https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923#issuecomment-1121950702
- [x] `go-md2man` | Check if `gh` is successfully built.
- [ ] `rclone` | Skipped as its unit test requires compile without LTO.
- [ ] `xray` https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923#issuecomment-1148588022 | Functionality tests are required as this upgrade enables `-buildmode=pie`
- [ ] `v2ray` https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923#issuecomment-1148577048 | Functionality tests are required as this upgrade enables `-buildmode=pie`
- [ ] `keybase` | May open a new topic to package other co
- [ ] `geoipupdate` | Functionality tests are required as this upgrade enables `-buildmode=pie`
- [ ] `ccat` https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923#issuecomment-1148603458
- [x] `hub` | Functionality tests are required as this upgrade enables `-buildmode=pie` | https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923#issuecomment-1121837159
- [ ] `kcptun` | Upstream does not provide unit test | Functionality tests are required as this upgrade enables `-buildmode=pie`
- [ ] ~~`flatbuffers`~~ | Skipped. Will be dropped out of go-rebuilds.
- [ ] `syncthing` https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923#issuecomment-1148621989
- [ ] `baidupcs-go` | Minor update. May not be tested.
- [x] `pcstat` | Switch buildtype to ABTYPE.
- [ ] `gogs` | https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923#issuecomment-1128624326 | Functionality tests are required as this upgrade enables `-buildmode=pie` 
- [ ] `direnv` | https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923#issuecomment-1148238124 | Functionality tests are required as this upgrade enables `-buildmode=pie`
- [ ] `gost` | Upstream not provide / Tests are broken | Functionality tests are required as this upgrade enables `-buildmode=pie`
- [ ] `android-platform-tools` | Minor update; Tests are not required.
- [ ] `elvish` | https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923#issuecomment-1148171959 | Major changes are applied. Requires tests.
- [ ] `pup` https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923#issuecomment-1148648223
- [ ] `fzf` https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923#issuecomment-1148639835
- [ ] `restic` https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923#issuecomment-1148662131
- [ ] `mongodb` | Build failed due to changes to glibc 2.34 |Database upgrade may require additional user configuration 
- [ ] `scons` | Builddep of MongoDB 4.4.14 . Not a go package.
3. Packages to rebuild
- [ ] `libguestfs`: Too complex. Will not upgrade in this topic (or after all other main upgrade complete)
- [ ] `v2ray-plugins` | Functionality tests are required because this update switches build method to ABTYPE
- [ ] `obfs4proxy`
- [ ] `dep`
- [ ] `glide`
- [ ] `podman` | Upgrade of which may be proceeded in next OCI/container survey;

**Comment by CamberLoid** [comment]
Test checklist for `gotop` | Pass if normally executed.
(Reset as new patch introduced)
- [x] amd64
image
- [x] arm64
image
- [x] loongson3
image
- [x] ppc64el
image
- [x] riscv64
image

**Comment by CamberLoid** [comment]
Patches in commit [0a71c63](https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923/commits/0a71c63df0b405818070e4615b0b306270c64108) and [ca7dfe9](https://github.com/AOSC-Dev/aosc-os-abbs/pull/3923/commits/ca7dfe94787a551e26e3c454e530ea9da0dc543b) have been formed into a pull request and submitted to upstream. This patch will solve a FTBFS issue on riscv64 platform.

Reference: 
- https://github.com/xxxserxxx/gotop/pull/218
- https://github.com/xxxserxxx/gotop/pull/218/commits/658926bd0fa8e15c75c4db1426f2b20307312870

**Comment by CamberLoid** [comment]
Tests for clash(FOSS) are skipped because a checking dependency `docker` is missing, which is required by upstream-shipped test suite.

**Comment by CamberLoid** [comment]
Test for `hub` will be performed by adding these lines to `autobuild/beyond` (will be discarded in formal build). Check if no error reports.
```bash
abinfo "Checking ..."
pushd "$SRCDIR"/src/github.com/github/hub
    make test || abwarn "Test failed ..."
popd
```
- [x] amd64 | Log: https://paste.aosc.io/paste/2dY4lujv5PMWcKV7uOWDbg 
image
- [x] arm64
image
- [x] loongson3
image
- [x] ppc64el
image
- [x] riscv64
image

**Comment by CamberLoid** [comment]
Tests for `gh` will be performed using integrated check by adding testing to `autobuild/beyond` (will be restored). Check if build succeeds with no error.
cmdline:
```bash
(echo -e "\
abinfo \"Checking ...\"
(make test || abdie \"Test failed.\") && abinfo \"Test passed.\"
" >> extra-vcs/gh/autobuild/beyond ) && ciel clean && ciel build -i go-survey gh && git restore . && ciel clean && ciel build -i go-survey gh
```
- [x] amd64 (pushed)
- [x] arm64
- [x] loongson3
- [x] ppc64el
- [x] riscv64

**Comment by CamberLoid** [comment]
Unit tests for `flatbuffers` are skipped as built-in tests are broken (also reported by archlinux).

**Comment by CamberLoid** [comment]
`flatbuffers` does not require `go` and `openjdk` to build. Will be dropped out of `groups/go-rebuilds` by the end of survey.

File: extra-shells/elvish/autobuild/postinst [hunk 1]
```diff
@@ -0,0 +1,5 @@
+grep -Fqx '/bin/elvish' /etc/shells || \
+grep -Fqx '/usr/bin/elvish' /etc/shells || \
```
Comment (MingcongBai) [inline extra-shells/elvish/autobuild/postinst]: Try and add some indents after line breaks to denote a chain of commands.

```
grep -Fqx '/bin/elvish' /etc/shells || \
    grep -Fqx '/usr/bin/elvish' /etc/shells || \
    echo '/usr/bin/elvish' >> /etc/shells
```

No need to tell user about /etc/shells update.

File: extra-shells/elvish/autobuild/postrm [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# This command is adopted from Arch Linux's elvish.install
+# Reference: 
+# https://github.com/archlinux/svntogit-community/blob/packages/elvish/trunk/elvish.install
```
Comment (MingcongBai) [inline extra-shells/elvish/autobuild/postrm]: Use a specific commit snapshot for your reference.

File: extra-shells/elvish/autobuild/postrm [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+if [ $1 = 'remove' ]; then
+    echo "Attempting to purge elvish from /etc/shells ..."
```
Comment (MingcongBai) [inline extra-shells/elvish/autobuild/postrm]: Again, no need to add outputs like this. We might address this in a shell survey for consistency.

File: extra-shells/elvish/autobuild/prepare [hunk 4]
```diff
@@ -0,0 +1,6 @@
+abinfo "Setting GOFLAGS ..."
```
Comment (MingcongBai) [inline extra-shells/elvish/autobuild/prepare:1]: Be a bit more specific about what GOFLAGS you've set here, and move the all-arch export line directly after this line.

File: extra-shells/elvish/autobuild/build [hunk 5]
```diff
@@ -0,0 +1,26 @@
+export GOPATH="$SRCDIR/abgopath"
+
+# The following command is partitially adopted from Arch Linux's PKGBUILD
```
Comment (MingcongBai) [inline extra-shells/elvish/autobuild/build]: "Adapt" *, and also by adapting a script, whether it's partial or not is less relevant. Minor note, but do fix your word choice.

File: extra-shells/elvish/autobuild/build [hunk 6]
```diff
@@ -0,0 +1,26 @@
+export GOPATH="$SRCDIR/abgopath"
```
Comment (MingcongBai) [inline extra-shells/elvish/autobuild/build:2]: Missing abinfo.

File: extra-shells/elvish/autobuild/build [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    -o abbuild \
+    ./cmd/elvish
+
+abinfo "Installing files ..."
+install -Dvm755 -t "$PKGDIR"/usr/bin "$SRCDIR"/abbuild/elvish
+install -Dvm644 -t "$PKGDIR"/usr/share/doc/"$PKGNAME" \
+    "$SRCDIR"/{README,SECURITY}.md
+
+abinfo "Installing documentations ..."
+mkdir -vp "$PKGDIR"/usr/share/doc/elvish/website
```
Comment (MingcongBai) [inline extra-shells/elvish/autobuild/build]: Why "website"? You could consider renaming the README.md above to ELVISH.md.

**Review by MingcongBai — CHANGES_REQUESTED** [review]

**Comment by CamberLoid** [comment]
Test build for arm64 is successfully performed

**Comment by CamberLoid** [comment]
Tests for `gogs` is successfully performed on the following platforms: `gogs web` + `lynx localhost:3000`
- [x] `amd64`
image
- [x] `arm64`
image
- [ ] `loongson3`
- [ ] `riscv64`

**Comment by OriginCode** [comment]
image

**Comment by CamberLoid** [comment]
From this point, test will be performed via a patched version of autobuild3.

See also: 
- https://github.com/AOSC-Dev/autobuild3/pull/129

**Comment by CamberLoid** [comment]
Test checklist for `elvish`: 
- [x] `amd64`
image
- [x] `arm64`
image
- [ ] `loongson3`
- [ ] `riscv64`

Script: https://github.com/AOSC-Dev/aosc-os-abbs/commit/c69c4222f3ab7d0cec23c4cc59cae7ec0a40cdfb

**Comment by CamberLoid** [comment]
Test checklist for `direnv`: 
- [x] `amd64` | Test exited with acceptable known failure
image
- [x] `arm64`
image
- [ ] `loongson3`
- [ ] `riscv64`

Script: https://github.com/AOSC-Dev/aosc-os-abbs/commit/35e457be4a8f25db5ee7f3182d922b611a13e49e

**Comment by CamberLoid** [comment]
Test checklist for `git-lfs`:
- [x] `amd64` no major failure
image
- [x] `arm64` 
image
- [ ] `loongson3`
- [ ] `riscv64`

**Comment by CamberLoid** [comment]
Unit test for `rclone` is skipped as it requires rclone to be built without LTO, which violates out package building principles.

**Comment by CamberLoid** [comment]
pcstat: LGTM on amd64
image

**Comment by CamberLoid** [comment]
v2ray: test checklist
- [x] `amd64` with a minor failure on TestTCPLocalNameServer
image
- [x] `arm64`
image
- [ ] `loongson3`
- [ ] `riscv64`

**Comment by CamberLoid** [comment]
xray:
- [x] `amd64` minor failure as geoip is not available, consider passed\
image
- [x] `arm64`
- [ ] `loongson3`
- [ ] `riscv64`

**Comment by CamberLoid** [comment]
ccat:
- [x] `amd64`: seems good to me, at least it works
image
- [ ] `arm64`
- [ ] `loongson3`
- [ ] `riscv64`

**Comment by CamberLoid** [comment]
syncthing: 
- [x] `amd64`
image
- [x] `arm64`
image
- [ ] `loongson3`
- [ ] `riscv64`

**Comment by CamberLoid** [comment]
fzf:
- [x] `amd64`
image
- [x] `arm64`
- [ ] `loongson3`
- [ ] `riscv64`

**Comment by CamberLoid** [comment]
pup: with an expected difference since https://github.com/ericchiang/pup/commit/6d1318bfa484b60309fe7a13e036093c179f15d6
- [x] `amd64`
image
- [x] `arm64`
image
- [ ] `loongson3`
- [ ] `riscv64`

**Comment by CamberLoid** [comment]
restic: unit test failed as fusermount failed inside container. May need further test on non-container environment.
image
- [ ] `amd64`
- [ ] `arm64`
- [ ] `loongson3`
- [ ] `riscv64`

**Comment by CamberLoid** [comment]
Test for restic will be skipped and placed to next survey.
Marking this PR as ready for review

**Review by MingcongBai — DISMISSED** [review]

**Review by MingcongBai — DISMISSED** [review]

**Review by MingcongBai — DISMISSED** [review]

**Review by MingcongBai — APPROVED** [review]

**Pull request merged**